QueryDatabaseAuthenticationHandler not using passwordEncoder

- [5.0.0] CAS server version
- Embedded Tomcat on OSx
- Configure cas.authn.jdbc.query[0].passwordEncoder.type=BCRYPT and use DatabaseAuthentication query. Authentication is not successful due code in **QueryDatabaseAuthenticationHandler** class

` if(!StringUtils.equals(password, e))` because that is not right way to check if there is match. Instead passwordEncoder.matches should be used where first argument is raw password as provided by user. Reason is that BCryptPasswordEncoder will not encode same password to same string so comparing them is irrelevant.

i am not yet familiar with gradle and master branch has issues

```
FAILURE: Build failed with an exception.

* What went wrong:
Could not resolve all dependencies for configuration ':core:cas-server-core-configuration-cloud-vault:compileClasspath'.
> A conflict was found between the following modules:
   - org.apache.httpcomponents:httpcore:4.4.5
   - org.apache.httpcomponents:httpcore:4.4.4

```

is it something I am doing wrong or it is really an issue there?

--- a/core/cas-server-core-authentication/src/main/java/org/apereo/cas/authentication/handler/support/AbstractUsernamePasswordAuthenticationHandler.java
+++ b/core/cas-server-core-authentication/src/main/java/org/apereo/cas/authentication/handler/support/AbstractUsernamePasswordAuthenticationHandler.java
@@ -77,7 +77,7 @@ public abstract class AbstractUsernamePasswordAuthenticationHandler extends Abst
      * @throws GeneralSecurityException On authentication failure.
      * @throws PreventedException On the indeterminate case when authentication is prevented.
      */
-    protected HandlerResult authenticateUsernamePasswordInternal(UsernamePasswordCredential transformedCredential,final String originalPassword)
+    protected HandlerResult authenticateUsernamePasswordInternal(final UsernamePasswordCredential transformedCredential,final String originalPassword)
             throws   GeneralSecurityException, PreventedException {
         return authenticateUsernamePasswordInternal(transformedCredential);
     }

--- a/core/cas-server-core-authentication/src/main/java/org/apereo/cas/authentication/handler/support/AbstractUsernamePasswordAuthenticationHandler.java
+++ b/core/cas-server-core-authentication/src/main/java/org/apereo/cas/authentication/handler/support/AbstractUsernamePasswordAuthenticationHandler.java
@@ -9,6 +9,7 @@ import org.apereo.cas.authentication.PreventedException;
 import org.apereo.cas.authentication.UsernamePasswordCredential;
 import org.apereo.cas.authentication.handler.PrincipalNameTransformer;
 import org.apereo.cas.authentication.support.PasswordPolicyConfiguration;
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
 import org.springframework.security.crypto.password.NoOpPasswordEncoder;
 import org.springframework.security.crypto.password.PasswordEncoder;
 
@@ -60,7 +61,25 @@ public abstract class AbstractUsernamePasswordAuthenticationHandler extends Abst
         userPass.setUsername(transformedUsername);
         userPass.setPassword(this.passwordEncoder.encode(userPass.getPassword()));
 
-        return authenticateUsernamePasswordInternal(userPass);
+        return authenticateUsernamePasswordInternal(userPass,((UsernamePasswordCredential) credential).getPassword());
+    }
+
+    /**
+     * Authenticates a username/password credential by an arbitrary strategy with extra parameter original credential password before
+     * encoding password. Override it if implementation need to use original password for authentication.
+     *
+     * @param transformedCredential the credential object bearing the transformed username and password.
+     * @param originalPassword original password from credential before password encoding
+     *
+     * @return HandlerResult resolved from credential on authentication success or null if no principal could be resolved
+     * from the credential.
+     *
+     * @throws GeneralSecurityException On authentication failure.
+     * @throws PreventedException On the indeterminate case when authentication is prevented.
+     */
+    protected HandlerResult authenticateUsernamePasswordInternal(UsernamePasswordCredential transformedCredential,final String originalPassword)
+            throws   GeneralSecurityException, PreventedException {
+        return authenticateUsernamePasswordInternal(transformedCredential);
     }
 
     /**
@@ -107,4 +126,19 @@ public abstract class AbstractUsernamePasswordAuthenticationHandler extends Abst
         }
         return false;
     }
+
+    /**
+     * Used in case passwordEncoder is used to match raw password with encoded password. Mainly for BCRYPT password encoders where each encoded
+     * password is different and we cannot use traditional compare of encoded strings to check if passwords match
+     *
+     * @param charSequence raw not encoded password
+     * @param password encoded password to compare with
+     *
+     * @return true in case charSequence matched encoded password
+     */
+    protected boolean matches(final CharSequence charSequence,final String password){
+        return this.passwordEncoder.matches(charSequence,password);
+    }
+
+
 }

--- a/core/cas-server-core-authentication/src/main/java/org/apereo/cas/authentication/handler/support/AbstractUsernamePasswordAuthenticationHandler.java
+++ b/core/cas-server-core-authentication/src/main/java/org/apereo/cas/authentication/handler/support/AbstractUsernamePasswordAuthenticationHandler.java
@@ -9,7 +9,6 @@ import org.apereo.cas.authentication.PreventedException;
 import org.apereo.cas.authentication.UsernamePasswordCredential;
 import org.apereo.cas.authentication.handler.PrincipalNameTransformer;
 import org.apereo.cas.authentication.support.PasswordPolicyConfiguration;
-import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
 import org.springframework.security.crypto.password.NoOpPasswordEncoder;
 import org.springframework.security.crypto.password.PasswordEncoder;
 

--- a/support/cas-server-support-jdbc/src/main/java/org/apereo/cas/adaptors/jdbc/QueryDatabaseAuthenticationHandler.java
+++ b/support/cas-server-support-jdbc/src/main/java/org/apereo/cas/adaptors/jdbc/QueryDatabaseAuthenticationHandler.java
@@ -29,7 +29,7 @@ public class QueryDatabaseAuthenticationHandler extends AbstractJdbcUsernamePass
     private String sql;
 
     @Override
-    protected HandlerResult authenticateUsernamePasswordInternal(final UsernamePasswordCredential credential)
+    protected HandlerResult authenticateUsernamePasswordInternal(final UsernamePasswordCredential credential,final String originalPassword)
             throws GeneralSecurityException, PreventedException {
 
         if (StringUtils.isBlank(this.sql) || getJdbcTemplate() == null) {
@@ -41,7 +41,9 @@ public class QueryDatabaseAuthenticationHandler extends AbstractJdbcUsernamePass
         final String password = credential.getPassword();
         try {
             final String dbPassword = getJdbcTemplate().queryForObject(this.sql, String.class, username);
-            if (!StringUtils.equals(password, dbPassword)) {
+
+            if ((StringUtils.isNotBlank(originalPassword) && !this.matches(originalPassword, dbPassword)) ||
+                    (StringUtils.isBlank(originalPassword) && !StringUtils.equals(password, dbPassword))) {
                 throw new FailedLoginException("Password does not match value on record.");
             }
         } catch (final IncorrectResultSizeDataAccessException e) {
@@ -53,6 +55,14 @@ public class QueryDatabaseAuthenticationHandler extends AbstractJdbcUsernamePass
             throw new PreventedException("SQL exception while executing query for " + username, e);
         }
         return createHandlerResult(credential, this.principalFactory.createPrincipal(username), null);
+    }
+
+
+    @Override
+    protected HandlerResult authenticateUsernamePasswordInternal(final UsernamePasswordCredential credential)
+                throws GeneralSecurityException, PreventedException
+    {
+        return authenticateUsernamePasswordInternal(credential,null);
     }
 
     public void setSql(final String sql) {

--- a/support/cas-server-support-jdbc/src/test/java/org/apereo/cas/adaptors/jdbc/QueryDatabaseAuthenticationHandlerTests.java
+++ b/support/cas-server-support-jdbc/src/test/java/org/apereo/cas/adaptors/jdbc/QueryDatabaseAuthenticationHandlerTests.java
@@ -10,6 +10,7 @@ import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.beans.factory.annotation.Qualifier;
 import org.springframework.boot.test.context.SpringBootTest;
 import org.springframework.cloud.autoconfigure.RefreshAutoConfiguration;
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
 import org.springframework.test.context.ContextConfiguration;
 import org.springframework.test.context.junit4.SpringRunner;
 
@@ -20,6 +21,8 @@ import javax.persistence.Id;
 import javax.security.auth.login.AccountNotFoundException;
 import javax.security.auth.login.FailedLoginException;
 import javax.sql.DataSource;
+import java.nio.charset.StandardCharsets;
+import java.security.SecureRandom;
 import java.sql.Connection;
 import java.sql.Statement;
 
@@ -89,7 +92,7 @@ public class QueryDatabaseAuthenticationHandlerTests {
         q.setDataSource(this.dataSource);
         q.setSql(SQL);
         q.authenticateUsernamePasswordInternal(
-                CoreAuthenticationTestUtils.getCredentialsWithDifferentUsernameAndPassword("usernotfound", "psw1"));
+                CoreAuthenticationTestUtils.getCredentialsWithDifferentUsernameAndPassword("usernotfound", "psw1"), "psw1");
 
     }
 
@@ -99,7 +102,7 @@ public class QueryDatabaseAuthenticationHandlerTests {
         q.setDataSource(this.dataSource);
         q.setSql(SQL);
         q.authenticateUsernamePasswordInternal(
-                CoreAuthenticationTestUtils.getCredentialsWithDifferentUsernameAndPassword("user1", "psw11"));
+                CoreAuthenticationTestUtils.getCredentialsWithDifferentUsernameAndPassword("user1", "psw11"), "psw11");
 
     }
 
@@ -109,7 +112,7 @@ public class QueryDatabaseAuthenticationHandlerTests {
         q.setDataSource(this.dataSource);
         q.setSql(SQL);
         q.authenticateUsernamePasswordInternal(
-                CoreAuthenticationTestUtils.getCredentialsWithDifferentUsernameAndPassword("user0", "psw0"));
+                CoreAuthenticationTestUtils.getCredentialsWithDifferentUsernameAndPassword("user0", "psw0"), "psw0");
 
     }
 
@@ -119,7 +122,7 @@ public class QueryDatabaseAuthenticationHandlerTests {
         q.setDataSource(this.dataSource);
         q.setSql(SQL.replace("password", "*"));
         q.authenticateUsernamePasswordInternal(
-                CoreAuthenticationTestUtils.getCredentialsWithDifferentUsernameAndPassword("user0", "psw0"));
+                CoreAuthenticationTestUtils.getCredentialsWithDifferentUsernameAndPassword("user0", "psw0"), "psw0");
 
     }
 
@@ -130,5 +133,43 @@ public class QueryDatabaseAuthenticationHandlerTests {
         assertNotNull(q.authenticateUsernamePasswordInternal(
                 CoreAuthenticationTestUtils.getCredentialsWithDifferentUsernameAndPassword("user3", "psw3")));
 
+    }
+
+    /**
+     *  This test proves that in case BCRYPT is used authentication using encoded password always fail
+     * with FailedLoginException
+     * @throws Exception
+     */
+    @Test(expected = FailedLoginException.class)
+    public void verifyBCryptFail() throws Exception {
+        final QueryDatabaseAuthenticationHandler q = new QueryDatabaseAuthenticationHandler();
+        q.setDataSource(this.dataSource);
+
+        BCryptPasswordEncoder encoder = new BCryptPasswordEncoder(8,
+                new SecureRandom("secret".getBytes(StandardCharsets.UTF_8)));
+
+        q.setSql(SQL.replace("password", "'" + encoder.encode("psw0") +"' password"));
+
+        q.setPasswordEncoder(encoder);
+        q.authenticateUsernamePasswordInternal(
+                CoreAuthenticationTestUtils.getCredentialsWithDifferentUsernameAndPassword("user0", "psw0"));
+    }
+
+    /**
+     *  This test proves that in case BCRYPT and using raw password test can authenticate
+     */
+    @Test
+    public void verifyBCryptSuccess() throws Exception {
+        final QueryDatabaseAuthenticationHandler q = new QueryDatabaseAuthenticationHandler();
+        q.setDataSource(this.dataSource);
+
+        BCryptPasswordEncoder encoder = new BCryptPasswordEncoder(6,
+                new SecureRandom("secret2".getBytes(StandardCharsets.UTF_8)));
+
+        q.setSql(SQL.replace("password", "'" + encoder.encode("psw0") +"' password"));
+
+        q.setPasswordEncoder(encoder);
+        q.authenticateUsernamePasswordInternal(
+                CoreAuthenticationTestUtils.getCredentialsWithDifferentUsernameAndPassword("user3", "psw0"),"psw0");
     }
 }

--- a/support/cas-server-support-jdbc/src/test/java/org/apereo/cas/adaptors/jdbc/QueryDatabaseAuthenticationHandlerTests.java
+++ b/support/cas-server-support-jdbc/src/test/java/org/apereo/cas/adaptors/jdbc/QueryDatabaseAuthenticationHandlerTests.java
@@ -148,11 +148,11 @@ public class QueryDatabaseAuthenticationHandlerTests {
         BCryptPasswordEncoder encoder = new BCryptPasswordEncoder(8,
                 new SecureRandom("secret".getBytes(StandardCharsets.UTF_8)));
 
-        q.setSql(SQL.replace("password", "'" + encoder.encode("psw0") +"' password"));
+        q.setSql(SQL.replace("password", "'" + encoder.encode("pswbc1") +"' password"));
 
         q.setPasswordEncoder(encoder);
         q.authenticateUsernamePasswordInternal(
-                CoreAuthenticationTestUtils.getCredentialsWithDifferentUsernameAndPassword("user0", "psw0"));
+                CoreAuthenticationTestUtils.getCredentialsWithDifferentUsernameAndPassword("user0", "pswbc1"));
     }
 
     /**
@@ -166,10 +166,11 @@ public class QueryDatabaseAuthenticationHandlerTests {
         BCryptPasswordEncoder encoder = new BCryptPasswordEncoder(6,
                 new SecureRandom("secret2".getBytes(StandardCharsets.UTF_8)));
 
-        q.setSql(SQL.replace("password", "'" + encoder.encode("psw0") +"' password"));
+        q.setSql(SQL.replace("password", "'" + encoder.encode("pswbc2") +"' password"));
 
         q.setPasswordEncoder(encoder);
-        q.authenticateUsernamePasswordInternal(
-                CoreAuthenticationTestUtils.getCredentialsWithDifferentUsernameAndPassword("user3", "psw0"),"psw0");
+        assertNotNull(q.authenticateUsernamePasswordInternal(
+                CoreAuthenticationTestUtils.getCredentialsWithDifferentUsernameAndPassword("user3", "pswbc2"),"pswbc2"));
+
     }
 }

